internal/cr_validation_test.go

```diff
@@ -0,0 +1,393 @@
+//
+// DISCLAIMER
+//
+// Copyright 2023 ArangoDB GmbH, Cologne, Germany
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+
+package internal
+
+import (
+	"encoding/json"
+	"fmt"
+	"go/ast"
+	"go/token"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	appsv1 "github.com/arangodb/kube-arangodb/pkg/apis/apps/v1"
+	backupv1 "github.com/arangodb/kube-arangodb/pkg/apis/backup/v1"
+	deploymentv1 "github.com/arangodb/kube-arangodb/pkg/apis/deployment/v1"
+	deploymentv2alpha1 "github.com/arangodb/kube-arangodb/pkg/apis/deployment/v2alpha1"
+	replicationv1 "github.com/arangodb/kube-arangodb/pkg/apis/replication/v1"
+	replicationv2alpha1 "github.com/arangodb/kube-arangodb/pkg/apis/replication/v2alpha1"
+	storagev1alpha "github.com/arangodb/kube-arangodb/pkg/apis/storage/v1alpha"
+	"github.com/arangodb/kube-arangodb/pkg/util"
+)
+
+func (def DocDefinition) ApplyToSchema(docName string, s *apiextensions.JSONSchemaProps) {
+	for _, e := range def.Enum {
+		z := strings.Split(e, "|")
+		s.Enum = append(s.Enum, apiextensions.JSON{
+			Raw: []byte("\"" + z[0] + "\""),
+		})
+	}
+
+	if def.Immutable != nil {
+		s.XValidations = append(s.XValidations, apiextensions.ValidationRule{
+			Rule:    fmt.Sprintf("self%s == oldSelf%s", def.Path, def.Path),
+			Message: fmt.Sprintf("field %s is immutable", strings.TrimPrefix(def.Path, ".")),
+		})
+	}
+
+	if len(docName) > 0 {
+		s.ExternalDocs = &apiextensions.ExternalDocumentation{
+			URL: getDocsLinkForField(docName, def),
+		}
+	}
+	s.Description = strings.Join(def.Docs, "\n")
+}
+
+func Test_GenerateCRValidationSchemas(t *testing.T) {
```

Comment:
We should add a comment here that we used that test to generate schema files

Reply:
done